File: Chris/2022-07-13-CourseSchedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        # Create prereqMap
        prereqMap = {i:[] for i in range(numCourses)}
        
        
        for course, prereq in prerequisites:
            prereqMap[course].append(prereq)
        
        # Set of visited vertices in current DFS  
        visitSet = set()
        
        def dfs(course):
            if course in visitSet:
                return False
            
            if prereqMap[course] == []:
                return True
            
            visitSet.add(course)
            for prereq in prereqMap[course]:
                if not dfs(prereq):
                    return False
            visitSet.remove(course)
            prereqMap[course] = []    
            return True
        
        for c in range(numCourses):
            if not dfs(c):
                return False
        return True

File: Chloe Gwon/2022-07-13-Course-Schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    let map = {};
    for (let i=0; i<numCourses; i++){
        map[i] = [];
    }
    
    let visited = {};
    for (let i=0; i<prerequisites.length; i++){
        map[prerequisites[i][0]].push(prerequisites[i][1]);
    }
    
    let dfs = function(prerequisites){
        if(visited[prerequisites]) return false;
        if (map[prerequisites].length==0) return true;
        visited[prerequisites] = true;
        for(let i=0; i<map[prerequisites].length; i++){
            if(!dfs(map[prerequisites][i])) return false;
        }
        delete visited[prerequisites]
        map[prerequisites] = [];
        return true;
    }
    for (let i=0; i<numCourses; i++){
        if(!dfs(i)) return false
    }
    return true;
};
